### PR TITLE
Add stdout "buffering" to slurm

### PR DESCRIPTION
**Pull request by ronawho**
For single locale performance testing we have a couple tests that generate
massive output files (fasta, a writeln test, and a few others.)

For batch mode, slurm buffers stdout on a per line basis and syncs the stdout
buffer to a file on the login node every so often. This constant syncing really
slows down these programs. For instance fasta generates a ~250 MB file and
previously this test took ~100 seconds to run. By sending output to /tmp and
then copying it, the tests only takes ~4 seconds (roughly the same time it
takes on chap04!, and the same time it takes for interactive mode)

This buffering works by setting slurm's --output to /tmp/<jobid>.out and then
after the job is completed moving the /tmp file to the output file that a user
specified or slurm auto-generated. In my (limited) testing, this also works for
multi-locale tests since all output is piped through a single node.

There's a pretty fat comment in in the slurm-srun launcher that goes into some
more detail.

I'm not sure if there's a similar issue for tests with a large stdin, but if
there is we can use slurms sbcast to take care of it.

This also updates chpl_launchcmd to use the stdout buffer for single locale
performance testing. I think it could set it in general, but there's no need to
since that's the only configuration with tests that produce massive amounts of
output

### CODE REVIEW

**Comment by ronawho** [comment]
@thomasvandoren can you review? Using brad's scale, I think this is a 7/7 in terms of pride :)  Well maybe not code pride, but I didn't think I was going to find a workaround to this issue.

File: runtime/src/launch/slurm-srun/launch-slurm-srun.c [hunk 1]
```diff
@@ -224,14 +249,30 @@ static char* chpl_launch_create_command(int argc, char* argv[],
...
     else {
-      fprintf(slurmFile, "#SBATCH --output=%s.%%j.out\n", argv[0]);
+      sprintf(stdoutFile,      "%s.%s.out", argv[0], "%j");
+      sprintf(stdoutFileNoFmt, "%s.%s.out", argv[0], "$SLURM_JOB_ID");
+    }
+
+    // If we're buffering the output, set the temp output file name.
+    // It's always /tmp/binaryName.<jobID>.out.
+    if (bufferStdout != NULL) {
+      sprintf(tmpStdoutFile,      "/tmp/%s.%s.out", argv[0], "%j");
```
Comment (thomasvandoren) [inline runtime/src/launch/slurm-srun/launch-slurm-srun.c]: I think it's usually a better practice to use `$TMPDIR` var, then `P_tmpdir` macro, and default to `/tmp`. I know we use `/tmp` in most places, though, so no need to do that here. Just thought I'd mention it.

**Comment by thomasvandoren** [comment]
Looks good :+1:

File: runtime/src/launch/slurm-srun/launch-slurm-srun.c (hunk 1 above)
Comment (ronawho) [inline runtime/src/launch/slurm-srun/launch-slurm-srun.c]: I'm not sure I'm following: Would this just become:
' "sprintf(tmpStdoutFile,      "%s/%s.%s.out", argv[0], P_tmpdir, "%j")  ' 

then, or is there more to it?

Comment (ronawho) [inline runtime/src/launch/slurm-srun/launch-slurm-srun.c]: Or is it:

// check if $TMPDIR is in the env, if so use it
// if it's not try to use P_tempdir if it's defined
// and fallback to /tmp if those two aren't available?

Comment (ronawho) [inline runtime/src/launch/slurm-srun/launch-slurm-srun.c]: I think I have something. I'll commit this for now, and add the updated version some other time.

**Pull request merged**